File: test/Interop/SwiftToCxx/functions/Inputs/throwing-noexcept.cpp
#include "throwing.h"
#include <utility>

using ThrowingNoexcept::Reference;
using ThrowingNoexcept::Value;

static_assert(!noexcept(ThrowingNoexcept::freeFunction()), "throwing function");
static_assert(!noexcept(Value::init(0)), "throwing initializer");
static_assert(!noexcept(std::declval<const Value &>().read()),
              "throwing method");
static_assert(!noexcept(std::declval<Value &>().update()), "throwing mutator");
static_assert(!noexcept(Value::make()), "throwing static method");
static_assert(!noexcept(std::declval<const Value &>().getComputed()),
              "throwing getter");
static_assert(!noexcept(std::declval<const Value &>().isPositive()),
              "throwing Bool getter");
static_assert(!noexcept(Value::getAnswer()), "throwing static getter");
static_assert(!noexcept(std::declval<const Value &>()[0]),
              "throwing subscript");
static_assert(!noexcept(std::declval<Reference &>().getComputed()),
              "throwing class getter");
static_assert(!noexcept(std::declval<const Reference &>()[0]),
              "throwing class subscript");

File: test/Interop/SwiftToCxx/functions/throwing-noexcept.swift
// RUN: %empty-directory(%t)
// RUN: %target-swift-frontend %s -module-name ThrowingNoexcept -clang-header-expose-decls=all-public -enable-experimental-feature GenerateBindingsForThrowingFunctionsInCXX -typecheck -emit-clang-header-path %t/throwing.h
// RUN: %target-interop-build-clangxx -std=c++17 -fsyntax-only %S/Inputs/throwing-noexcept.cpp -I %t -DSWIFT_CXX_INTEROP_EXPERIMENTAL_SWIFT_ERROR
// REQUIRES: swift_feature_GenerateBindingsForThrowingFunctionsInCXX

public func freeFunction() throws {}
public struct Value {
  public let number: Int
  public init(_ number: Int) throws { self.number = number }
  public func read() throws -> Int { number }
  public mutating func update() throws {}
  public static func make() throws -> Int { 42 }
  public var computed: Int { get throws { number } }
  public var isPositive: Bool { get throws { number > 0 } }
  public static var answer: Int { get throws { 42 } }
  public subscript(index: Int) -> Int { get throws { number + index } }
}

public class Reference {
  public init() {}
  public var computed: Int { get throws { 0 } }
  public subscript(index: Int) -> Int { get throws { index } }
}
